### ISSUE TITLE
[question] Retain reordered column order

### ISSUE THREAD

**Issue by DeanFromLondon** [mui/mui-x#5704]
### Order ID 💳

48444

### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### The problem in depth 🔍

I am trying to retain the new reordered columns between refreshes by saving it to LocalStorage.
This should allow the user to save it's customised view of the DataGrid table.

I went through the documentation and searched through issues here and in stackoverflow. It appears that the only way to set the order of the columns is by changing their default order when setting the `columns` prop.

As an attempt to walk-around this issue, I tried saving the newly changed index of columns after every column change (using `onColumnOrderChange`) but didn't succeed (caused re-rendering issues and faulty column moves functionality when attempted to push `GridColDef` objects into `Columns:GridColDef[]` array.).

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  System:
    OS: macOS 12.4
  Binaries:
    Node: 17.9.0 - ~/.nvm/versions/node/v17.9.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v17.9.0/bin/yarn
    npm: 8.5.5 - ~/.nvm/versions/node/v17.9.0/bin/npm
  Browsers:
    Chrome: 104.0.5112.79 - Tested
    Edge: Not Found 
    Firefox: 103.0.1 - Tested
    Safari: 15.5 - Tested
```
</details>

**Comment by cherniavskii** [comment]
Hey @DeanFromLondon 
Did you try [state import / export](https://mui.com/x/react-data-grid/state/#restore-the-state-with-apiref)?
Sound like it should work for your use case

**Comment by DeanFromLondon** [comment]
Thank you for your reply!
I am managing several states manually (such as pagination, server side filtering, server side sorting) and exporting/importing their values. I will try to combine them with exporting the state and importing only the column order.
Is there an option to control column order with props, same way as I control the attributes mentioned above?

**Comment by cherniavskii** [comment]
> Is there an option to control column order with props, same way as I control the attributes mentioned above?

Not really, but controlled column ordering might be a good addition to the API.
Similar to [controlled column pinning](https://mui.com/x/react-data-grid/column-pinning/#controlling-the-pinned-columns), it would add two props - one with ordered column fields, and a callback that would be called when user changes columns order.

**Comment by msykora** [comment]
I have a question related to this, since I see some other "duplicates" closed, mentioned it's expected behavior and you should memoize columns and am not sure I can expect this fixed.
I think it's not good behavior, if you include orderedFields in initialState when they don't actually work and are immediately rewritten on refresh (shown in https://github.com/mui/mui-x/issues/5418). I have a component based on grid pro where I memoize the columns passed in a prop and then persist orderedFields in localStorage to persist column order, and it doesn't work.
To work around this I guess I'll have to manually reorder and the re-memoize the columns prop on every column order change on top of persisting orderedFields, do you think that's "as expected"? That initialState prop orderedFields does not actually override the column order, and when it's different from columns prop order, it show the initialState order for plsit second, then refreshes with column prop order?
I hope not and I hope this will be fixed, but from reading all comments in this and related issues, I don't have certainty.

**Comment by DeanFromLondon** [comment]
Hello again @cherniavskii,
Sorry for the delayed response. This matter had been in the backlog for some time but I am finally able to address it.

The state API's import/export methods partially solved my issue. Exporting and then importing the DataGrid's state only works on one of two table instances. Let's name them instance A (working) and instance B (not working) for readability. 

Both still print out a warning:

1. `Warning: Failed prop type: The prop apiRef.current is marked as required in ForwardRef(DataGridPro), but its value is `null`.
2. `Warning: Failed prop type: Invalid prop initialState of type string supplied to ForwardRef(DataGridPro), expected object.`

On instance A, which is working, I am using the following props:

- `selectionModel` & `onSelectionModelChange`
- `columnVisibilityModel` & `onColumnVisibilityModelChange`
- `rowsPerPageOptions`
- `pageSize` & `onPageSizeChange`
- `sortModel` & `onSortModelChange`
- `filterModel` & `onFilterModelChange`
- `onRowClick`
- `apiRef`
- `initialState`
- `sx`
- `onColumnOrderChange`
- `onColumnResize`

On instance B, which is a bit more complex, i'm using the following props:

- `apiRef`
- `initialState`
- `sx`
- `loading`
- `filterMode="server"`
- `onFilterModelChange`
- `filterModel`
- `sortingMode="server"`
- `onSortModelChange`
- `sortModel`
- `pagination={true}`
- `paginationMode="server"`
- `onPageSizeChange`
- `page`
- `pageSize`
- `rowsPerPageOptions`
- `onPageChange`
- `rowCount`
- `onRowClick`
- `columnVisibilityModel`
- `onColumnVisibilityModelChange`
- `sortingOrder={['asc', 'desc']}`
- `onColumnOrderChange`
- `onColumnResize`
-```
 componentsProps={{
          // Disable OR filter
          filterPanel: { linkOperators: ['and'] },
          // Custom right click menu
          row: {
            onContextMenu: handleContextMenu,
            style: { cursor: 'context-menu' },
          }
        }}
```

The issue I am experiencing with instance B is the columns are bouncing around. Disabling `onColumnOrderChange` stops them from bouncing around and allows me to reorder the columns, but then of course I cannot save the state for later use.
In instance A, the same functionality works and both are simply calling the method:
`const exportState = () => { setState(apiRef.current.exportState()); }`

> > Is there an option to control column order with props, same way as I control the attributes mentioned above?

> Not really, but controlled column ordering might be a good addition to the API. Similar to [controlled column pinning](https://mui.com/x/react-data-grid/column-pinning/#controlling-the-pinned-columns), it would add two props - one with ordered column fields, and a callback that would be called when user changes columns order.

This sounds like a great solution for the problem and it is just what I had in mind opening this issue. 
Please let me know if I should upvote (👍) this topic anywhere else.

**Comment by cherniavskii** [comment]
Hey @DeanFromLondon 

> The issue I am experiencing with instance B is the columns are bouncing around

This sounds like a bug. Can you share a minimal reproduction example?

> Disabling onColumnOrderChange stops them from bouncing around and allows me to reorder the columns, but then of course I cannot save the state for later use.

If you're using `onColumnOrderChange` only to persist column ordering state, you should not need it since `exportState` should also export column order state.

**Comment by DeanFromLondon** [comment]
@cherniavskii Thank you for your quick reply!

I have managed to solve the issue described yesterday by changing the columns from a regular `const` to a React state.
I have updated the comment thinking you would be able to address the latest issue but you were replying just as I was editing it :)

The tables still issue a warning as described in my last comment.

> If you're using onColumnOrderChange only to persist column ordering state, you should not need it since exportState should also export column order state.

I am using `onColumnOrderChange` to invoke the `exportState` method. 
Would you suggest using any other event to detect changes in the columns order?

**Comment by cherniavskii** [comment]
@DeanFromLondon 
> I have managed to solve the issue described yesterday by changing the columns from a regular const to a React state.

Yeah, columns have to be memoized, otherwise column order will reset on the next render. You can also use `React.useMemo` instead of state (unless you need state to update columns)

> I am using onColumnOrderChange to invoke the exportState method.

Makes sense then :+1:

**Comment by TamirCode** [comment]
> > Is there an option to control column order with props, same way as I control the attributes mentioned above?
> 
> Not really, but controlled column ordering might be a good addition to the API. Similar to [controlled column pinning](https://mui.com/x/react-data-grid/column-pinning/#controlling-the-pinned-columns), it would add two props - one with ordered column fields, and a callback that would be called when user changes columns order.

i need a way to set column order (so it would be controlled)
and initialState is broken as well https://github.com/mui/mui-x/issues/6206

we have a `onColumnOrderChange` prop, now we just need a `columnOrder` prop, just like we have `onColumnVisibilityModelChange` and `columnVisibilityModel`

will also be included in apiRef.current.setColumnOrder

so definitely implement this essential feature, every time my dynamic data changes it resets their column order

**Comment by hafley66** [comment]
It would be really nice to have a consistent api, `columnOrderModel` and `onColumnOrderModelChange`. The issue is that `onColumnOrderChange` fires _while_ the user is dragging across other columns, not after they let go of the drag.

I am saving datagrid state with an api, not local storage, so i can't use `onColumnOrderModelChange` to call exportState and save it, not even with debouncing. This has been a frustrating experience, as i have been saving filter/sort/pagination models to our api's to remember ui grid state. I have extensive wrapper around the data-grid and am trying to add controlled column order, but its an incredibly error prone process if i dont want to call my api too much

**Comment by dennykoch** [comment]
I'm having the same problem right now. exportState gave me problems in the past (not every state i needed was included and reinitialising felt inconsistent), so I switched to wiring every single part of the state myself. Now I'am stuck because columnOrder is not a fully controlled state. I could implement it myself by manipulating colDefs, but this introduces a kind of "chicken and egg" problem where I first need all the colDefs and than would have to reorder them. This feels not very clean.

**Comment by konrazem** [comment]
I have custom dialog to restore columns settings saved in localStorage. I am looking for method like 
`current.setColumnsOrderedField(config.columns.orderedFields)`. I cannot use method current.setState

**Comment by itsjaked** [comment]
I agree this is important. Our implementation is getting a bit crazy having to hack together the state management. We allow users to create columns, change the initial state, etc. Everything is saved after they take the action (a separate "Save State" button is too clunky, but would make implementation somewhat easier).

**Comment by skye-levit** [comment]
Any updates on the issue?

**Comment by ben-gryps** [comment]
I am having a similar issue. I would like to control the column order in the grid through an external component and have them both be in sync.

**Comment by destegabry** [comment]
I also need to control the order of the columns, or at least be able to rely on `initialState.columns.orderedFields` (which seems ingored by `@mui/x-data-grid-premium": "^7.1.0"`

**Comment by mparsakia** [comment]
The datagrid ignores, or does not process properly the orderedFields when the initial state contains pinnedColumns.left or right.

**Comment by markookuka** [comment]
+1 on the introduction of a `columnOrderModel`

**Comment by segullshairbutt** [comment]
The issue was opened in 2022 and is still not progressing in 2024. 😄

**Comment by edwardhuntatco** [comment]
This might not be the solution for everyone, but it seems to be working for me. 

Context - Mui Data Grid Pro. I have enabled the pinned rows, and column ordering. Everything works fine until a user pins a row and it sets the column order back to the original order in grid col def. The pinned row Ids in the actions cell are a dependency of the column definition useMemo, which is what triggers the column re-creation.

This issue is also visible in the MUI docs themselves. 

https://mui.com/x/react-data-grid/row-pinning/

Navigate to the controlled pinned rows demo, change the column order, and then adjust the pinned rows. It will reset the column order.

My solution is to add the following code to my global styled data grid implementation. Essentially setting the order fields back to my controlled value if the pinned rows change. This solution has been implemented and is functioning in a prod environment.

`
  useEffect(() => {
      gridApiRef?.current?.setState((state) => ({
        ...state,
        columns: { ...state.columns, orderedFields },
      }));

  }, [pinnedRows]);
`

**Comment by willcolbert** [comment]
using an initialState prop in the DataGrid component of type GridInitialState with  {columns: { orderedFields: [ORDER]}} where ORDER is a string[] of the field names worked for me

**Comment by jkunwar** [comment]
I had the same issue. I was restoring the state from localstorage. using timeout for setting the orderFields worked for me

` if (apiRef?.current?.restoreState) {
        apiRef.current.restoreState(state)
        setTimeout(() =>
          apiRef.current.restoreState({
            columns: { orderedFields: state.columns?.orderedFields },
          }),
        )
      }
    },`

**Comment by luisfuertes** [comment]
initialState doesn't work for me either; the table ignores it. It ignores columns.dimensions and columns.orderedFields. Not being able to use these properties in a controlled way is a pretty big problem. I use the version `"@mui/x-data-grid-pro": "^6.19.2"`

This is that i finally did:
```
    const formattedColumns = useMemo(() => {
      // SET COLUMNS WIDTH
      const columnsData = props.columns.map(column => {
        const width = initialState?.columns?.dimensions?.[column.field]?.width || column.width || undefined

        return {
          ...column,
          width,
          flex: undefined,
        }
      })

      // SORT COLUMNS BY CONFIG DATA ORDERED FIELDS
      const orderedFields = initialState?.columns?.orderedFields || []
      columnsData?.sort((a, b) => {
        const indexA = orderedFields.indexOf(a.field)
        const indexB = orderedFields.indexOf(b.field)
        return indexA - indexB
      })

      return columnsData
    }, [initialState, props.columns])
```

**Comment by mtsmith** [comment]
restoreState will ignore orderedFields and use current state.columns.orderedFields instead.

Maybe resetting state.columns.orderedFields before calling restoreState will help?

**Comment by ngelotte** [comment]
Is there any plan to introduce columnOrderModel or something like it?
Currently it is very hard to restore column order or custom column width to facilitate saving of the grid state and restoring it.
